File: daily_python/1653_Minimum_Deletions_to_Make_String_Balanced.py
import os
import sys
from typing import List

class Solution:
    def minimumDeletions(self, s: str) -> int:
        acount = s.count('a')
        ans = min(len(s) - acount, acount)
        bcount = 0
        for c in s:
            if c == 'a':
                acount -= 1
            else:
                bcount += 1
            ans = min(ans, acount + bcount)
        return ans

def ReadFileData(filename: str):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        lines = content.split('\n')
        for line in lines:
            data = line.split('=')[1].strip().strip('\";')
            dataset.append(data)
    return dataset


if __name__ == '__main__':
    if (len(sys.argv) == 1):
        pathname = os.path.basename(__file__)
        filename = os.path.splitext(pathname)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = ReadFileData(filename)
    solution = Solution()
    results = []
    for data in dataset:
        results.append(solution.minimumDeletions(data))
    for index, result in enumerate(results):
        print(f'data {index+1} : {result}')

File: daily/2024/1653_Minimum_Deletions_to_Make_String_Balanced.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    int minimumDeletions(string s) {
        int n = s.length();
        int a_count = 0;
        for (auto c : s) if (c == 'a') a_count++;
        int ans = min(a_count, n - a_count);
        int b_count = 0;
        for (int critical_point = 0; critical_point < n; critical_point++){
            if (s[critical_point] == 'a') a_count--;
            else b_count++;
            ans = min(ans, b_count + a_count);
        }
        return ans;
    }
};

int main(){
    string s = "aababbab";
    // string s = "bbaaaaabb";
    Solution S;
    cout << S.minimumDeletions(s) << endl;
    return 0;
}